Need a short description (OOPS!).

chgibb-apple opened the pull request:
#### 16a15e2b4283dbea29f3a4cf9b6f4fd41b34c72e
<pre>
Need a short description (OOPS!).
Need the bug URL (OOPS!).

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/delete-if-too-large:
</pre>

aproskuryakov commented on Tools/CISupport/Shared/maybe-delete-directory.bash:
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
Nit: we'd normally use `.sh` extension for the file. the shebang looks fine.

Slightly more substantial, please use `/bin/sh`. The macOS version of `bash` is very old, and actually prints out warnings about pending removal if I remember correctly.

We'd probably use Python for something like this, but I think that shell is OK.

chgibb-apple replied:
`bash` on macOS prints a deprecation warning when invoked interactively but does not when invoked non-interactively (such as by `bash foo.bash`). This scripts uses a number of bash-isms that do not work in `sh`. This script does appear to work in `zsh` however. I've changed the shebang and file extension to `zsh`.

aproskuryakov replied:
No, please don't use zsh either. We standardized on /bin/sh shebang intentionally.

chgibb-apple replied:
Okay. I will rewrite this script to remove the bash-isms that `sh` does not support.

aproskuryakov commented on Tools/CISupport/Shared/maybe-delete-directory.bash:
```diff
@@ -0,0 +1,45 @@
+
+if [[ $# < 1 ]]
+then
+    echo "Must provide absolute path to directory to prune."
```
Nit: it would be nicer to print this out as "Usage" help,  as opposed to having the computer give orders to humans :)

aproskuryakov commented on Tools/CISupport/Shared/maybe-delete-directory.bash:
```diff
@@ -0,0 +1,45 @@
+
+if [[ -L "${1%/}" ]]
+then
+    echo "$1 must not be a symbolic link."
```
Similarly, I'd have the script say that it `$1` is a symlink, please provide a direct directory link (not quite sure why symlinks aren't OK though).

aproskuryakov commented on Tools/CISupport/build-webkit-org/factories.py:
```diff
@@ -38,6 +38,7 @@ def __init__(self, platform, configuration, architectures, buildOnly, additional
             self.addStep(KillOldProcesses())
         self.addStep(CleanBuildIfScheduled())
         self.addStep(DeleteStaleBuildFiles())
+        self.addStep(MaybePruneCoreSymbolicationdCache())
```
Not sure if "Maybe" adds valuable information to the step name. How is it more conditional than existing steps?

aproskuryakov replied:
Also, I think that this script would be executed on non-Mac queues too (Windows and Linux). We probably want to make it conditional on Mac being the host platform (keeping iOS Simulator in mind).

chgibb-apple replied:
A "Maybe" prefix is a naming pattern I've used in the past to describe an operation that may (or may not) cause an observable side effect. I've renamed this class to make it more clear.

I've also added a condition to this step to only run on Darwin hosts.

roynos replied:
We've used "IfNecessary" suffix.

aproskuryakov commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -203,6 +203,13 @@ class KillOldProcesses(shell.Compile):
+
+class MaybePruneCoreSymbolicationdCache(shell.Compile):
+    name = "maybe-prune-coresymbolicationd-cache"
+    description = ["pruning coresymbolicationd cache if it is over 10GB in size"]
```
"pruning coresymbolicationd cache to < 10GB" is shorter wording for your consideration.

aproskuryakov reviewed (commented):
Added some minor comments. Probably best for someone other than me to do final review.

aproskuryakov commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -203,6 +203,13 @@ class KillOldProcesses(shell.Compile):
+    name = "maybe-prune-coresymbolicationd-cache"
+    description = ["pruning coresymbolicationd cache if it is over 10GB in size"]
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    command = ["bash", "Tools/CISupport/Shared/maybe-delete-directory.bash", "/System/Library/Caches/com.apple.coresymbolicationd"]
```
With the shebang and executable bit set, this can be just `command = ["Tools/CISupport/Shared/maybe-delete-directory.bash", ...`.

JonWBedard commented on Tools/CISupport/Shared/maybe-delete-directory.bash:
```diff
@@ -0,0 +1,45 @@
+ awk ' { last = $0; } END { print last; } ' | \
+ awk ' { print $1; }' | \
+ awk ' { if ( $1 > 10737418240) { print "true"; } else { print "false"; } } ' && \
+ printf '\0')
```
I don't think I'm going to be able to follow through on reviewing this, so this should be taken as a possible suggestion rather than request.

One of the things @aj062 and I have done with some more complicated buildbot steps is use `||` in a shell command to conditionally run a single command in a buildbot step without a stand-alone script. It results in the active code being quite ugly, but it's also nice because it lets us have a step that's detached from the state of the checkout and works regardless of the branch we're running on. This particular bit of code is guaranteed to have a checkout, so that isn't strictly necessary, but also, we have single check followed by a single command that is only relevant on Mac, which fits the shell-command from buildbot pattern.

chgibb-apple replied:
Can you point me towards some prior art demonstrating what you've described please? I'm not sure I understand the pattern you've described.

JonWBedard replied:
In `Tools/CISupport/ews-build/steps.py`, if you search `self.shell_command`, you can see the places where we're running ugly compound commands. The reason I call this possibility out is that the code as proposed requires us to cherry-pick this change to a branch, probably two branches, actually. Which is certainly a thing we can do, it just increases the cost of a separate file in a buildbot step like this.

chgibb-apple replied:
Using [`CheckOutSpecificRevision`](https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L678) and the buildbot documentation as an example, I think that structure mostly makes sense. Is there a clean way to run a specific step locally? Assuming I would like to rewrite `PruneCoreSymbolicationdCacheIfTooLarge` as a `shell.Command` and run just that locally, how would I go about doing that?

With respect to unit testing, is there a preferred way to unit test `shell.Command` subclasses?

JonWBedard replied:
I think you want to try and break down that script into 2 commands and join them with `&&`, the first being the check and the second being the removal. I would make sure that the combinations run locally in your terminal. Next step is running a test on our staging instance, ews-build.webkit-uat.org. That's a real buildbot instance which a few machines attached. You want to have it do a build that runs your change (cc @aj062 to help). Last step would be creating a unit test, which is mostly just going to be a string check. `CheckOutSpecificRevision` has unit tests in in `steps_unittest.py`, my suggestion is you copy and mutate those because that step has relatively simple tests.

chgibb-apple replied:
I've rewritten all the logic in Python instead of stretching to use shell. I tried to use `getProperty`/`setProperty` to inject fakes for `Path` and some other functionality to try to make this testable. It appears this does not work because properties have to be JSON serialisable.

roynos commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +205,13 @@ class KillOldProcesses(shell.Compile):
+    name = "prune-coresymbolicationd-cache-if-too-large"
+    description = ["pruning coresymbolicationd cache to < 10GB"]
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    command = ["zsh", "Tools/CISupport/Shared/maybe-delete-directory.zsh", "/System/Library/Caches/com.apple.coresymbolicationd"]
```
where is `maybe-delete-directory.zsh`? Is it a separate patch?

roynos replied:
I don't know what the script does. I would delete the sub directories. Instead of deleting the folder itself.

roynos replied:
have you considered using `find` with `atime` to target the directories that no one is accessing. this should prevent applications from crashing because files and folders no longer exist.

chgibb-apple replied:
This was rewritten as Python.

roynos commented on Tools/CISupport/Shared/maybe-delete-directory.sh:
```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
```
most of our multiline scripts are written in Python. I am not opposed to landing this script as a bash script.

chgibb-apple replied:
This was rewritten as Python.

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
+
+    def start(self):
+        directory = '/System/Library/Caches/com.apple.coresymbolicationd'
+        directoryPath = Path(directory).resolve()
```
Does this actually operate on the worker? cc @aj062 to double check that If so, I _love_ the approach!

aproskuryakov replied:
Indeed, let's confirm that! I suspect that it's probably running on the master.

aj062 replied:
Yes, I think it would run on master.
For example this email code (`send_email_for_git_issue`) in `start()` also run on server: https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L658

JonWBedard commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
     command = ["python3", "Tools/CISupport/kill-old-processes", "buildbot"]
 
 
+class PruneCoreSymbolicationdCacheIfTooLarge(buildstep.BuildStep):
```
~Generally, this is only relevant for Mac queues, right?~

JonWBedard replied:
~Might want to add a `doStepIf`~

JonWBedard replied:
Looks like you did that in the factory

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
+    name = "prune-coresymbolicationd-cache-if-too-large"
+    description = ["pruning coresymbolicationd cache to < 10GB"]
+    descriptionDone = ["pruned coresymbolicationd cache"]
+    flunkOnFailure = True
```
Do we really want to fail the build (and have red ews status-bubbles) if this step fails? 
Seems like this and haltOnFailure should be no.

chgibb-apple replied:
Both have been changed to `False`.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +206,38 @@ class KillOldProcesses(shell.Compile):
+            self.finished(FAILURE)
+            return None
+
+        if directoryPath.stat().st_size >= 10737418240:
```
Better to make this number a variable with readable name, and also have it explicitly as a multiplication to indicate what it represent.

chgibb-apple replied:
Done.

aj062 reviewed (changes requested):
r- since this would run the commands on master and not on the bots. It might not be straight-forward to do the directory size checks on worker and delete the directory on worker in ShellCommands. I'm fine with the previous script approach (adding a shell/python script which would run on worker) as that would be more flexible and easily extendable in future. 

Can also make this script generic to take care of any kind of deletion (and on multiple platforms), for example it can even be extended in future to include the fix for <rdar://problem/72849252>.

aj062 commented on Tools/CISupport/build-webkit-org/steps.py:
```diff
@@ -205,6 +206,39 @@ class KillOldProcesses(shell.Compile):
+            self.finished(FAILURE)
+            return None
+
+        tenGigabytes = 1024 * 1000 * 1000 * 10
```
Nit: I guess you meant 1024 instead of 1000

chgibb-apple replied:
Fixed.

aj062 commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,25 @@
+import sys
```
shebang missing

chgibb-apple replied:
Fixed.

aj062 commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,25 @@
+    tenGigabytes = 1024 * 1024 * 1024 * 10
+    if directoryPath.stat().st_size >= tenGigabytes:
+        os.remove(directory)
+        print("Successfully removed " + directory)
```
Maybe better to say: f"Deleted {directory}" (instead of using the word successfully)

chgibb-apple replied:
Done.

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+    directoryPath = Path(directory).resolve()
+
+    if not directoryPath.exists():
+        print(directory + "does not exist")
```
this will `directory_namedoes not exist` I recommend using `.format()` or `f strings`. since this is python3 specific you should use strings.

chgibb-apple replied:
Fixed. Thank you!

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+
+    if not directoryPath.is_dir():
+        print(directory + "is not a directory")
+        return 1
```
nit: I would change the return value here. you can check against the return value without parsing standard output.

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+    if directoryPath.stat().st_size >= tenGigabytes:
+        os.remove(directory)
+        print("Deleted " + directory)
+
```
`return 0` not sure what sys.exit(None) resolves as.

chgibb-apple replied:
Fixed. Thank you!

roynos replied:
Fixed in a new PR?

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+
+
+def main():
+    directory = sys.argv[1]
```
use `argparse` ?

chgibb-apple replied:
What does `argparse` provide over using `sys.argv`?

roynos replied:
it will make it easier to add arguments, description, and help output.

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+        return 1
+
+    tenGigabytes = 1024 * 1024 * 1024 * 10
+    if directoryPath.stat().st_size >= tenGigabytes:
```
I would make this argument and call it something like `size_threshold`. I am only saying this because the name is generic `delete-if-too-large`

roynos commented on Tools/Scripts/delete-if-too-large:
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python3
```
not sure how to check this via GitHub. but did you make this script executable?

aj062 replied:
Please add Copyright message (you can copy it from another similar file).

chgibb-apple committed:
Prune coresymbolicationd cache if it grows larger than 10G
rdar://92755815

Reviewed by Aakash Jain.

* Tools/CISupport/build-webkit-org/factories.py:
(Factory.__init__):
* Tools/CISupport/build-webkit-org/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/build-webkit-org/steps.py:
(KillOldProcesses):
(MaybePruneCoreSymbolicationdCache):

Canonical link: https://commits.webkit.org/253061@main
